plugins/org.eclipse.n4js.dts/src/org/eclipse/n4js/dts/astbuilders/DtsNamespaceBuilder.java

```diff
@@ -69,14 +107,84 @@ public DtsNamespaceBuilder(DtsTokenStream tokenStream, LazyLinkingResource resou
 				RULE_exportStatementTail);
 	}
 
+	/**
+	 * Overloads {@link #consume(ParserRuleContext)} to codify the rule that when a {@link NamespaceDeclarationContext}
+	 * is passed in, you will always get back a {@link N4NamespaceDeclaration} (never a {@link N4ModuleDeclaration}).
+	 */
+	public N4NamespaceDeclaration consume(NamespaceDeclarationContext ctx) {
+		return (N4NamespaceDeclaration) super.consume(ctx);
+	}
+
+	/**
+	 * Overloads {@link #consume(ParserRuleContext)} to codify the rule that when a {@link ModuleDeclarationContext} is
+	 * passed in, you might get back an {@link N4NamespaceDeclaration} or an {@link N4ModuleDeclaration}.
+	 */
+	public N4AbstractNamespaceDeclaration consume(ModuleDeclarationContext ctx) {
```

Comment:
I had separate sub-classes and merged them, because it isn't possible to get rid of the checks, anyway. However, I just double-checked this and found a solution that allows to get rid of at least some of the instanceof/checks.